draft/0002-subdocapi.md

```diff
@@ -0,0 +1,351 @@
+  parsing a single document.
+* DocumentTooDeep: Document is too deep to parse. 
+* ValueTooDeep: Proposed value would make the document too deep to parse
+
+### Numeric Errors
+These errors are only received for `counter_in` operations:
+
+* NumberTooBig: Existing number value in document is too big (out of range between `INT64_MIN` and `INT64_MAX`
+* ZeroDelta: Got a zero value for a delta
+* BadNumber: Value provided as delta cannot be parsed as an integer. This error should
+  usually not be seen with high level SDKs
+* DeltaTooBig: The delta itself is too big, or the result of the operation would yield
+  a number too big.
+
+
+# Language Specifics
+In this section, the abstract design parts need to be broken down on the SDK level by each maintainer and signed off eventually.
+
+## C
+
+The C SDK will not use these verbs, and will instead model the subdoc operations
+after the existing `lcb_store3()` API - using `lcb_sdstore3`:
+
+    lcb_CMDSDSTORE cmd = { 0 };
+    LCB_CMD_SET_KEY(&cmd, "key", 3);
+    LCB_CMD_SET_VALUE(&cmd, "\"value\"", strlen("\"value\""));
+    LCB_CMD_SET_VALUE(&cmd, "path", strlen("path"));
+    cmd.mode = LCB_SUBDOC_DICT_UPSERT;
+    lcb_sched_enter(instance);
+    lcb_sdstore3(instance, NULL, &cmd);
+    lcb_sched_leave(instance);
+    
+    
+## Python
+
+    from subdocument import get, exists, upsert, counter
+
+    cb.get_upsert_in('user:mnunberg', 'address', ['123 Main St', 'Reno', 'NV', 'USA'])
+    
+    result = cb.get_in('user:mnunberg', 'address[0]')
+    result.value
+    # '123 Main St'
+    
+    # Demonstrate lookup specs
+    from couchbase.subdocument import get, exists
+    results = cb.lookup_in('user:mnunberg',
+                           get('email'),
+                           get('address'),
+                           exists('couchbase_id'))
+    email_ok, email = results[0]
+    addr_ok, addr = results[1]
+    couchbase_exists, _ = results[2]
+    
+    # Demonstrate mutation specs
+    from couchbase.subdocument import extend, replace, arrayinsert, addunique
+    cb.mutate_in('user:mnunberg',
+                 extend('messages', {'from': 'management', 'body': 'Hello!'}, direction=FRONT),
+                 replace('email', 'mnunberg2000@juno.com'),
+                 arrayinsert('interests[4]', 'sitting'),
+                 addunique('likes', 'running'))
+                 
+
+
+| Generic | .NET         | Java   | NodeJS | Go | C | PHP | Python | Ruby |
+| ------- | ------------ | ------ | ------ | -- | - | --- | ------ | ---- |
+| foo()   | foo<dynamic> | Foo<D> | ...... | .. | . | ... | ...... | .... |
+
+
+## Notes on method naming
+
+Overloading classic KV operations to use an additional _path_ parameter is
+discouraged due to fundamental differences between `get(key)` and `get(key, path)`.
+The latter is not simply a modifier of the former (making the `get` more restricted)
+but rather an entirely different operation.
+
```

Comment:
Considering I don't deal much with languages which contain overloads, I'd say that this should be left to whatever is idiomatic within the language. Do bear in mind however that using overloads would also make the list of overloads for `get` very big!